crates/proto/src/xfer/dnssec_dns_handle/nsec3_validation.rs

```diff
@@ -512,20 +514,97 @@ fn validate_nodata_response(
+                .nsec3_data
+                .type_bit_maps()
+                .contains(&RecordType::CNAME)
         {
             return Proof::Bogus;
         } else {
             return Proof::Secure;
         }
     }
 
+    // Case 3:
+    // Query type is DS, records for name exist, but there are no DS records (opt-out proof)
+    //
+    // RFC 5155 § 6
+    //
+    //   In this specification, as in [RFC4033], [RFC4034] and [RFC4035], NS
+    //   RRSets at delegation points are not signed and may be accompanied by
+    //   a DS RRSet.  With the Opt-Out bit clear, the security status of the
+    //   child zone is determined by the presence or absence of this DS RRSet,
+    //   cryptographically proven by the signed NSEC3 RR at the hashed owner
+    //   name of the delegation.  Setting the Opt-Out flag modifies this by
+    //   allowing insecure delegations to exist within the signed zone without
+    //   a corresponding NSEC3 RR at the hashed owner name of the delegation.
+    //
+    //   An Opt-Out NSEC3 RR is said to cover a delegation if the hash of the
+    //   owner name or "next closer" name of the delegation is between the
+    //   owner name of the NSEC3 RR and the next hashed owner name.
+    //
+    //   An Opt-Out NSEC3 RR does not assert the existence or non-existence of
+    //   the insecure delegations that it may cover.  This allows for the
+    //   addition or removal of these delegations without recalculating or re-
+    //   signing RRs in the NSEC3 RR chain.  However, Opt-Out NSEC3 RRs do
+    //   assert the (non)existence of other, authoritative RRSets.
+    //
+    //   An Opt-Out NSEC3 RR MAY have the same original owner name as an
+    //   insecure delegation.  In this case, the delegation is proven insecure
+    //   by the lack of a DS bit in the type map and the signed NSEC3 RR does
+    //   assert the existence of the delegation.
+    //
+    //   Zones using Opt-Out MAY contain a mixture of Opt-Out NSEC3 RRs and
+    //   non-Opt-Out NSEC3 RRs.  If an NSEC3 RR is not Opt-Out, there MUST NOT
+    //   be any hashed owner names of insecure delegations (nor any other RRs)
+    //   between it and the name indicated by the next hashed owner name in
+    //   the NSEC3 RDATA.  If it is Opt-Out, it MUST only cover hashed owner
+    //   names or hashed "next closer" names of insecure delegations.
+    //
+    //   The effects of the Opt-Out flag on signing, serving, and validating
+    //   responses are covered in following sections.
+    //
+    // *Note*: the case of an opt-out NSEC3 record having the same original owner
+    // name as the hashed query name and not having the DS bit set in the type flags
+    // is covered here by case 2.
+    if query_type == RecordType::DS
+        && find_covering_record(nsec3s, &hashed_query_name, &base32_hashed_query_name[..])
+            .iter()
+            .all(|x| x.nsec3_data.type_bit_maps().contains(&query_type) && x.nsec3_data.opt_out())
```

Comment:
Nit: maybe `query_type` -> `RecordType::DS` here since that should already be true?